### PR TITLE
vm: adjust comparable limit for ByteArray

### LINKED ISSUE #2460: Limit EQUAL for structs
See neo-project/neo-vm#454.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2460.

### CODE REVIEW

File: pkg/vm/stackitem/item.go [hunk 1]
```diff
@@ -594,19 +606,33 @@ func (i ByteArray) TryInteger() (*big.Int, error) {
...
+	val, ok := s.(*ByteArray)
+	if !ok {
+		return false
+	}
+	comparedSize = lCurr
+	lOther := uint(len(*val))
+	if lOther > comparedSize {
+		comparedSize = lOther
+	}
 	if i == s {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/stackitem/item.go:629]: Not sure, what do you mean. They are pointers, and I've left this check.

File: pkg/vm/vm_test.go [hunk 2]
```diff
@@ -809,6 +809,30 @@ func TestEQUAL(t *testing.T) {
 	t.Run("Buffer", getTestFuncForVM(prog, false, stackitem.NewBuffer([]byte{42}), stackitem.NewBuffer([]byte{42})))
 }
 
+func TestEQUALByteArrayWithLimit(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/vm_test.go:812]: Could you, please, review the tests? Not sure the second one is correct.

File: pkg/vm/stackitem/item.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/vm/stackitem/item.go:634]: `i *ByteArray` vs `s Item`

> A value x of non-interface type X and a value t of interface type T are comparable when values of type X are comparable and X [implements](https://go.dev/ref/spec#Implementing_an_interface) T. They are equal if t's dynamic type is identical to X and t's dynamic value is equal to x.

It all works, but we have `val` now which should be a bit more efficient.

File: pkg/vm/stackitem/item.go [hunk 3]
```diff
@@ -594,19 +606,35 @@ func (i ByteArray) TryInteger() (*big.Int, error) {
...
 	}
-	if i == s {
-		return true
-	} else if s == nil {
-		return false
+	var comparedSize = 1
+	defer func() { *limit -= comparedSize }()
+
+	if s == nil {
+		return i == nil
```
Comment (roman-khimov) [inline pkg/vm/stackitem/item.go:623]: How can `i` be `nil` after `len(*i)`?

File: pkg/vm/stackitem/item.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/stackitem/item.go]: Got it, fixed!

File: pkg/vm/vm_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply pkg/vm/vm_test.go:812]: Got it, did you mean that the second case should fail on `maxComparableSize == 0` condition?

File: pkg/vm/stackitem/item.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/vm/stackitem/item.go]: I've added  the check that `i` is not `nil` in the start, just in case.

**Review by roman-khimov — APPROVED** [review]

File: pkg/vm/vm_test.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/vm/vm_test.go:812]: Yes, that was the intention (`ByteArray != Buffer`)

**Pull request merged**